jwst/pipeline/calwebb_spec2.py

```diff
@@ -201,80 +206,23 @@ def process_exposure_product(
                 else:
                     raise RuntimeError('Cannot determine WCS.')
 
-        # Do background processing, if necessary
-        if exp_type in WFSS_TYPES or len(members_by_type['background']) > 0:
-
-            if exp_type in WFSS_TYPES:
-                bkg_list = []           # will be overwritten by the step
-            else:
-                bkg_list = members_by_type['background']
-
-            # Setup for saving
-            self.bkg_subtract.suffix = 'bsub'
-            if multi_int:
-                self.bkg_subtract.suffix = 'bsubints'
-
-            # Backwards compatibility
-            if self.save_bsub:
-                self.bkg_subtract.save_results = True
-
-            # Call the background subtraction step
-            input = self.bkg_subtract(input, bkg_list)
-
-        # Apply NIRSpec MSA imprint subtraction
-        # Technically there should be just one.
-        # We'll just get the first one found
-        imprint = members_by_type['imprint']
-        if exp_type in ['NRS_MSASPEC', 'NRS_IFU'] and \
-           len(imprint) > 0:
-            if len(imprint) > 1:
-                self.log.warning('Wrong number of imprint members')
-            imprint = imprint[0]
-            input = self.imprint_subtract(input, imprint)
-
-        # Apply NIRSpec MSA bad shutter flagging
-        if exp_type in ['NRS_MSASPEC', 'NRS_IFU']:
-            input = self.msa_flagging(input)
+        # Steps whose order is the same for all types of input.
+        input = self.bkg_subtract(input, members_by_type['background'])
+        input = self.imprint_subtract(input, members_by_type['imprint'])
+        input = self.msa_flagging(input)
 
         # The order of the next few steps is tricky, depending on mode:
         # WFSS/Grism data need flat_field before extract_2d, but other modes
         # need extract_2d first. Furthermore, NIRSpec MOS and FS need
         # srctype and wavecorr before flat_field.
         if exp_type in ['NRC_WFSS', 'NIS_WFSS', 'NRC_TSGRISM']:
```

Comment:
Maybe use another macro here that includes all modes that use a grism, e.g. GRISM_TYPES? It consists of WFSS_TYPES + 'NRC_TSGRISM'.